Add an S3Path type

rofinn opened the pull request:
Blocked on tagging FilePathsBase.jl.

oxinabox commented on Project.toml:
```diff
@@ -0,0 +1,28 @@
+
+[compat]
+AWSCore = ">= 0.5"
+julia = "1"
```
Need compat version for FilePathsBase

rofinn replied:
That'll come when I tag a new release

oxinabox replied:
I don't understand

rofinn replied:
The compat will be the next version of FilePathsBase which I haven't released yet.

oxinabox commented on src/s3path.jl:
```diff
@@ -0,0 +1,235 @@
+    path = ()
+    dir = true
+
+    @assert startswith(str, "s3://")
```
This should be an `ArgumentError`

oxinabox commented on src/s3path.jl:
```diff
@@ -0,0 +1,235 @@
+    str = String(str)
+    root = ""
+    path = ()
+    dir = true
```
Can we call this `isdirectory`?
`dir` to me sounds like it might be the directory part of the name.

rofinn replied:
👌

oxinabox commented on src/s3path.jl:
```diff
@@ -0,0 +1,235 @@
+    if length(tokenized) > 3
+        root = "/"
+        # If the last tokenized element is an empty string then we've parsed a directory
+        dir = isempty(last(tokenized))
```
per above
```suggestion
        isdirectory = isempty(last(tokenized))
```

oxinabox commented on src/s3path.jl:
```diff
@@ -0,0 +1,235 @@
+        return fp.dir ? join(fp.segments, '/') * "/" : join(fp.segments, '/')
+    else
+        # Call getfield even though we know it'll error
+        # so the message is consistent.
```
Why don't we just delete the top?
```
  if isdefined(fp, attr)
        return getfield(fp, attr)
```

rofinn replied:
Hmmm, true. Ideally, since we control the type we shouldn't be too worried about overriding the field.

oxinabox commented on src/s3path.jl:
```diff
@@ -0,0 +1,235 @@
+
+# We need to special case join and parents so that we propagate
+# directories correctly
+function Base.join(prefix::S3Path, pieces::AbstractString...)
```
I love the fact that we get to do this so much.

the name `joinpath` was a clear indication of `join` begging for the use of dispatch on a path type.

oxinabox commented on src/s3path.jl:
```diff
@@ -0,0 +1,235 @@
+# directories correctly
+function Base.join(prefix::S3Path, pieces::AbstractString...)
+    segments = String[prefix.segments...]
+    dir = endswith(last(pieces), "/")
```
`isdirectory`

Is it worth abostracting this?
It is also called in the constryuctior

rofinn replied:
Maybe? We were initially doing this a bunch before the `dir`/`isdirectory` field was introduced. I'm fine to leave it with the 2 cases we're doing it, but we might want to update that in the future.

oxinabox commented on src/s3path.jl:
```diff
@@ -0,0 +1,235 @@
+            S3Path(fp.segments[1:i], fp.root, fp.drive, true, fp.config)
+        end
+    else
+        error("$fp has no parents")
```
Why doesn't this return itself.
after all `..` in the unix root directory leavies you in in the unix root directory

rofinn replied:
NOTE: This is also an issue with FilePathsBase. https://github.com/rofinn/FilePathsBase.jl/issues/19

oxinabox commented on src/s3path.jl:
```diff
@@ -0,0 +1,235 @@
+    return !isempty(objects)
+end
+
+Base.real(fp::S3Path) = fp
```
Can we comment this with reference to the FilePathBase docs?

rofinn replied:
Oops, we don't have pushed docs setup for FilePathsBase yet.

oxinabox commented on src/s3path.jl:
```diff
@@ -0,0 +1,235 @@
+            else
+                error(
+                    "The parent of $fp does not exist. " *
+                    "Pass recursive=true to create it."
```
```suggestion
                    "Pass `recursive=true` to create it."
```

oxinabox commented on src/s3path.jl:
```diff
@@ -0,0 +1,235 @@
+        end
+
+        write(fp, "")
+    end
```
Should return the newly created directory here.

oxinabox commented on src/s3path.jl:
```diff
@@ -0,0 +1,235 @@
+        basenames = Set(s3_get_name(k, string(o["Key"])) for o in objects)
+
+        # Lexographically sort the results
+        return sort!(filter!(!isempty, collect(basenames)))
```
Why do we construct a `Set` only to `collect` it?
if we mean `unique` then lets just say that?

oxinabox commented on src/s3path.jl:
```diff
@@ -0,0 +1,235 @@
+end
+
+function Base.readdir(fp::S3Path)
+    if isdir(fp)
```
Do we need
```suggestion
    if FilePathsBase.isdir(fp)
```
?
or is that the same as `Base.isdir`.
If it is, then I have mild objections to the earlier part where we overloiaded `FilePaths.isdir`
and it actually was changing e `Base.isdir`

rofinn replied:
It's the same as `Base.isdir`. If `FilePathsBase` is overloading `Base.isdir` does it matter?

oxinabox commented on src/s3path.jl:
```diff
@@ -0,0 +1,235 @@
+end
+
+Base.read(fp::S3Path) = s3_get(fp.config, fp.bucket, fp.key)
+Base.read(fp::S3Path, ::Type{String}) = String(read(fp))
```
We can probably get rid of this?
Since the new syntax is to hust call `String(read)`?

Or if we must keep this then we should just define this in `FilePathsBase`

rofinn replied:
I didn't realize `read(path, String)` was being deprecated.

oxinabox replied:
I think it is gone

rofinn replied:
👌

oxinabox commented on src/s3path.jl:
```diff
@@ -0,0 +1,235 @@
+function FilePathsBase.mktmpdir(parent::S3Path)
+    fp = parent / string(uuid4(), "/")
+    mkdir(fp)
+    return fp
```
Can delete this line if we fix `mkdir` to return the directory it mase

oxinabox commented on test/runtests.jl:
```diff
@@ -258,31 +260,32 @@ versions = s3_list_versions(aws, bucket_name, "key3")
 
 
-HTTP.ConnectionPool.showpool(stdout)
+# HTTP.ConnectionPool.showpool(stdout)
```
Why this all commented out?

oxinabox commented on test/s3path.jl:
```diff
@@ -0,0 +1,115 @@
+    ]
+
+    # Run all of the automated tests
+    test(ps, testsets)
```
Oh awesome.

nickrobinson251 commented on test/s3path.jl:
```diff
@@ -0,0 +1,115 @@
+
+function test_s3_mkdir(p::PathSet)
+    @testset "mkdir" begin
+        garply = p.root / "corge" / "grault" / "garply/"
```
some terrific naming here 😆

nickrobinson251 commented on test/s3path.jl:
```diff
@@ -0,0 +1,115 @@
+    # Run all of the automated tests
+    test(ps, testsets)
+
+    # TODO: Copy over specific tests that can't be tested reliably from the general case.
```
🆗

oxinabox commented on src/s3path.jl:
```diff
@@ -0,0 +1,235 @@
+        root = "/"
+        # If the last tokenized element is an empty string then we've parsed a directory
+        dir = isempty(last(tokenized))
+        path = tuple(filter!(!isempty, tokenized[4:end])...)
```
It depends on Julia version.
Annoyingly. In 1.0 `tuple(x...)`, in 1.1 `Tuple(x)`

If one really cares workout how to do it  uses `ntuple`.

omus replied:
@oxinabox what problem are you referring to? `Tuple` has always taken an iterator.

oxinabox replied:
@omus What is more performant between splatting + `tuple` and not splatting and `Tuple` varies between julia versions.
It is really weird but both Rory and I have done benchmarking on this.

rofinn replied:
If folks find the `Tuple` syntax cleaner I'm fine to switch it. I don't think we care too much about performance here.

rofinn commented on src/s3path.jl:
```diff
@@ -0,0 +1,235 @@
+function FilePathsBase.isdir(fp::S3Path)
+    # Note: objects "s3://bucket/a" and "s3://bucket/a/b" can co-exist. If both of these
+    # objects exist listing the keys for "s3://bucket/a" returns ["s3://bucket/a"] while
+    # "s3://bucket/a/" returns ["s3://bucket/a/b"].
```
Sure. I was going to leave a comment on the weirdness of directories on S3 in the constructor docstring.

rofinn replied:
It is not listed in the notes for the `S3Path` docstring.

rofinn commented on src/s3path.jl:
```diff
@@ -0,0 +1,235 @@
+# Need API for accessing object ACL permissions for this to work
+FilePathsBase.isexecutable(fp::S3Path) = false
+Base.isreadable(fp::S3Path) = true
+Base.iswritable(fp::S3Path) = true
```
Yes, and query the object ACL rules which isn't currently supported by AWSS3. I figured it would be better to work on that in a separate PR.

omus replied:
Can we avoid defining these methods until then?

rofinn replied:
I'd prefer to keep it under this comment as a reminder of which methods to update. I suppose we could comment it out, but I don't think it makes a big difference. It'll just error anyways.

rofinn commented on test/runtests.jl:
```diff
@@ -258,31 +260,32 @@ versions = s3_list_versions(aws, bucket_name, "key3")
+# ntasks=30)
 
-HTTP.ConnectionPool.showpool(stdout)
+# HTTP.ConnectionPool.showpool(stdout)
```
They seemed to cause tests to stall? I'll re-enable when I'm ready to remove the WIP, but for now it makes testing locally difficult.

rofinn replied:
Hmmm, I was unable to figure out why these tests were stalling, so I've left a comment explaining that they seem to stall. I can try figuring out what's happening, but that's unrelated to this PR.

omus replied:
Would have been better to use a comment block `#= =#` in this case

rofinn commented on Project.toml:
```diff
@@ -0,0 +1,28 @@
+
+[compat]
+AWSCore = ">= 0.5"
+julia = "1"
```
Compat for FilePathsBase has been added. I'm reluctant to add a new restriction for HTTP in this PR.

rofinn commented:
Alright, I think I've addressed most of the comments. I still don't understand why the unrelated asyncmap tests stall for me locally, but I let them run for 30 min and they never finished. I'm also unclear what they're supposed to be testing as object creation is handled elsewhere, so this is basically testing HTTP.jl?

omus commented on src/AWSS3.jl:
```diff
@@ -471,52 +474,63 @@ end
 `Owner`, `StorageClass`.
 """
-function s3_list_objects(aws::AWSConfig, bucket, path_prefix="")
+function s3_list_objects(aws::AWSConfig, bucket, path_prefix=""; delimiter="/", max_items=nothing)
```
Should the `delimiter` be a `Char`?

rofinn replied:
I don't think so because we're passing it as a string to S3 anyways.

omus commented on src/AWSS3.jl:
```diff
@@ -471,52 +474,63 @@ end
+            end
+            if marker != ""
+                q["marker"] = marker
+            end
```
Personal preference but this may be easier to read:
```julia
!isempty(path_prefix) && (q["prefix"] = path_prefix)
!isempty(delimiter) && (q["delimiter"] = delimiter)
!isempty(marker) && (q["marker"] = marker)
```

rofinn replied:
I'd tried to stick with the original convention in this function. I tend to agree with your preferred syntax, but I think we should change the convention everywhere in 1 PR.

omus commented on src/s3path.jl:
```diff
@@ -0,0 +1,278 @@
+    elseif attr === :bucket
+        return split(fp.drive, "//")[2]
+    elseif attr === :key
+        return fp.isdirectory ? join(fp.segments, '/') * "/" : join(fp.segments, '/')
```
```suggestion
        return join(fp.segments, '/') * (fp.isdirectory ? "/" : "")
```

omus commented on src/s3path.jl:
```diff
@@ -0,0 +1,278 @@
+    isdirectory = endswith(last(pieces), "/")
+
+    for p in pieces
+        push!(segments, filter!(!isempty, split(p, "/"))...)
```
```suggestion
        append!(segments, filter!(!isempty, split(p, "/")))
```

omus commented on src/s3path.jl:
```diff
@@ -0,0 +1,278 @@
+function Base.join(prefix::S3Path, pieces::AbstractString...)
+    isempty(pieces) && return prefix
+
+    segments = String[prefix.segments...]
```
```suggestion
    segments = Vector{String}(prefix.segments)
```

omus commented on src/s3path.jl:
```diff
@@ -0,0 +1,278 @@
+# are handled again.
+function FilePathsBase.sync(src::AbstractPath, dst::S3Path; delete=false)
+    # Create an index of all of the source files
+    index = Dict(Tuple(setdiff(p.segments, src.segments)) => p for p in walkpath(src))
```
Does this always work?
```julia
julia> setdiff(["a", "b", "c"], ["b", "a", "d"])
1-element Array{String,1}:
 "c"
```

rofinn replied:
Yes? We can trust that `src.segments` is always a direct subset of `p.segments` based on how `walkpath` works. Is there something else in your example that I'm supposed to be noticing?

omus replied:
Hmm, I see. Seems to be a slight abuse of `setdiff` but this should be ok

omus commented on src/s3path.jl:
```diff
@@ -0,0 +1,278 @@
+        objects = s3_list_objects(fp.config, fp.bucket, k; delimiter="")
+
+        # Only list the basename and not the full key
+        basenames = unique(s3_get_name(k, string(o["Key"])) for o in objects)
```
```suggestion
        basenames = unique!(s3_get_name(k, string(o["Key"])) for o in objects)
```

oxinabox replied:
Can't call `unique!` on a generator.

rofinn replied:
☝️

omus replied:
Don't use a generator?
```julia
julia> @btime unique(x for x in 1:100);
  2.394 μs (17 allocations: 6.17 KiB)

julia> @btime unique!([x for x in 1:100]);
  412.206 ns (1 allocation: 896 bytes)
```

oxinabox replied:
feels like worth openning an upstream issue on.

omus commented on test/s3path.jl:
```diff
@@ -0,0 +1,202 @@
+root = Path("s3://$bucket_name/pathset-root/")
+ps = PathSet(
+    root,
+    root / "foo/",
```
Just curious if `root / "foo" /` works?

omus replied:
Thinking about this more I realized we `/` to be supported as a post-fix operator.
